### ISSUE TITLE
libyang: No module connected with the prefix found

### ISSUE THREAD

**Issue by TeSwen** [CESNET/libyang#1975]
Hi.  When starting the FRR application I'm getting this error:  
2023/01/27 22:28:05 OSPF: [YEW9P-QECEK][EC 100663352] sysrepo: subscribing for configuration changes made in the 'frr-interface' module
2023/01/27 22:28:05 OSPF: [WKKKP-GJR2Z][EC 100663352] frr_sr_config_change_cb: SR_EV_CHANGE, clling prepare: 4
2023/01/27 22:28:05 OSPF: **libyang: No module connected with the prefix "frr-ospfd" found (prefix format JSON module names)**.
2023/01/27 22:28:05 OSPF: [V6124-SV3DH][EC 100663325] frr_sr_process_change: unknown data path: /frr-interface:lib/interface[name='tunl0101g0']/frr-ospfd:ospf/instance[id='1']

The data path is valid as can be demonstrated by performing a get-config through netopeer2-cli and passing that xpath on the running datastore,   Data is definitely returned.  Also a yang model for 'frr-ospfd' does exist in the yang repository.     What would it mean that no module with the prefix 'frr-ospfd' was found? 
The FRR has been compiled with these:  
libyang.so.2.25.4
libsysrepo.so.7.11.8

**Comment by jktjkt** [comment]
Based on the provided information, this doesn't look like a bug in libyang. How does the output of `sysrepoctl --list` look like?

**Comment by michalvasko** [comment]
>  Also a yang model for 'frr-ospfd' does exist in the yang repository.

Does this mean it is also **implemented**? It needs to be and if it is, you will find it in the output of `sysrepoctl -l`, what @jktjkt mentioned.

**Comment by TeSwen** [comment]
Yes the module is in sysrepo and I do have data associated with it in the datastore.   
sysrepoctl -l
[ec2-user@ip-10-148-33-154 ~]$ sudo /usr/local/bin/sysrepoctl -l
Sysrepo repository: /home/ec2-user/sysrepo/build/repository

Module Name                   | Revision   | Flags | Owner             | Startup Perms | Submodules | Features
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
frr-filter                    | 2019-07-04 | I     | root:root         | 666           |            |       
frr-interface                 | 2020-02-05 | I     | root:root         | 666           |            |       
frr-isisd                     | 2021-02-15 | I     | root:root         | 666           |            |       
frr-nexthop                   | 2019-08-15 | I     | root:root         | 666           |            |       
frr-ospf-route-map            | 2020-01-02 | I     | root:root         | 666           |            |       
**frr-ospfd**                     | 2020-07-21 | I     | root:root         | 666           |            |       
frr-pathd                     | 2018-11-06 | I     | root:root         | 666           |            |       
frr-ripd                      | 2019-09-09 | I     | root:root         | 666           |            |       
frr-ripngd                    | 2019-09-09 | I     | root:root         | 666           |            |       
...

Below is output from netopeer2-cli.   Shows that data does exist at the referenced xpath.    What does Michal mean by 'implemented'?   All data below is just sample/test data so did not mask off IPs. 

>get-config --source running --filter-xpath /frr-interface:lib/interface[name='tunl0101g0']/frr-ospfd:ospf/instance[id='1']
DATA
<data xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
  <lib xmlns="http://frrouting.org/yang/interface">
    <interface>
      <name>tunl0101g0</name>
      <ospf xmlns="http://frrouting.org/yang/ospfd">
        <instance>
          <id>1</id>
          <bfd>false</bfd>
          <network>point-to-point</network>
          <area>0</area>
          <authentication>
            <message-digest>false</message-digest>
            <null>false</null>
          </authentication>
          <message-digest-key>
            <key-id>6</key-id>
            <mds-key/>
          </message-digest-key>
          <authentication-key/>
          <cost>1</cost>
          <dead-interval>
            <interval>10</interval>
            <minimal>
              <hello-multiplier>1</hello-multiplier>
            </minimal>
          </dead-interval>
          <hello-interval>5</hello-interval>
          <retransmit-interval>3</retransmit-interval>
          <transmit-delay>2</transmit-delay>
          <mtu-ignore>false</mtu-ignore>
          <priority>1</priority>
          <interface-address>
            <address>192.168.0.10</address>
            <area>0</area>
            <authentication>
              <message-digest>false</message-digest>
              <null>false</null>
            </authentication>
            <message-digest-key>
              <key-id>7</key-id>
              <mds-key/>
            </message-digest-key>
            <authentication-key/>
            <cost>1</cost>
            <dead-interval>
              <interval>10</interval>
              <minimal>
                <hello-multiplier>1</hello-multiplier>
              </minimal>
            </dead-interval>
            <hello-interval>5</hello-interval>
            <retransmit-interval>3</retransmit-interval>
            <transmit-delay>2</transmit-delay>
            <mtu-ignore>false</mtu-ignore>
            <priority>1</priority>
          </interface-address>
        </instance>
      </ospf>
    </interface>
  </lib>
</data>

**Comment by michalvasko** [comment]
Okay, that looks fine, so what function exactly return the original error, with parameters?

**Comment by TeSwen** [comment]
Here is the snippet of code being called (returning NULL):
/*
 * Use libyang to find the schema node associated to the path and get
 * the northbound node from there (snode private pointer).
*/
	snode = lys_find_path(ly_native_ctx, NULL, path, 0);
	if (!snode)
		return NULL;

'path'  is "/frr-interface:lib/interface[name='tunl0101g0']/frr-ospfd:ospf/instance[id='1']

**Comment by michalvasko** [comment]
What exactly is `ly_native_ctx`, how did you get it?

**Comment by TeSwen** [comment]
More snippets. 

/* libyang container. */
struct ly_ctx *ly_native_ctx;


/* Initialize libyang container for native models. */
    ly_native_ctx = yang_ctx_new_setup(true, false);

struct ly_ctx *yang_ctx_new_setup(bool embedded_modules, bool explicit_compile)
{
	struct ly_ctx *ctx = NULL;
	const char *yang_models_path = YANG_MODELS_PATH;
	LY_ERR err;

	if (access(yang_models_path, R_OK | X_OK)) {
		yang_models_path = NULL;
		if (errno == ENOENT)
			zlog_info("yang model directory \"%s\" does not exist",
				  YANG_MODELS_PATH);
		else
			flog_err_sys(EC_LIB_LIBYANG,
				     "cannot access yang model directory \"%s\"",
				     YANG_MODELS_PATH);
	}

	uint options = LY_CTX_NO_YANGLIBRARY | LY_CTX_DISABLE_SEARCHDIR_CWD;
	if (explicit_compile)
		options |= LY_CTX_EXPLICIT_COMPILE;
	err = ly_ctx_new(yang_models_path, options, &ctx);
	if (err)
		return NULL;

	if (embedded_modules)
		ly_ctx_set_module_imp_clb(ctx, yang_module_imp_clb, NULL);

	return ctx;
}



/* Load YANG modules and their corresponding northbound callbacks. */
    for (size_t i = 0; i < nmodules; i++) {
        DEBUGD(&nb_dbg_events, "northbound: loading %s.yang",
               modules[i]->name);
        *loadedp++ = yang_module_load(modules[i]->name);
    }

**Comment by jktjkt** [comment]
Do I understand it correctly that this is a sysrepo callback (or something which is triggered via a sysrepo callback), but instead of using the libyang context which is available through sysrepo, the code creates a new libyang context from some on-disk state? That's a bug, just use the context which is available via the sysrepo session.

**Comment by TeSwen** [comment]
That's something I can try.

**Comment by michalvasko** [comment]
Please, just call `sr_session_acquire_context()` to get the *sysrepo* context, otherwise the state of `sysrepoctl -l` is completely irrelevant. But do not forget to call `sr_session_release_context()` once you are done with it.

**Comment by TeSwen** [comment]
I did change to use the sysrepo context using the above calls.    It did get rid if the original error 'libyang: No module connected with the prefix found' but it is still unable to find the node for my path.   So for this call, 

_snode = lys_find_path(sysrepo_ctx, NULL, path, 0);
if (!snode)
	return NULL;

return snode->priv;_

pretty sure its returning null because this value is null: _snode->priv_.  

Any more suggestions on what I can do?

**Comment by michalvasko** [comment]
What are you doing exactly? Could you please share all the details of your use-case, not just fragments? *libyang* has little to do with what is in the `priv` pointer, it is meant to be used by applications.

**Comment by TeSwen** [comment]
Before I get into the detail I had a question on the below subscription call.  
ret = sr_module_change_subscribe(
		session, module->name, NULL, frr_sr_config_change_cb, NULL, 0,
		SR_SUBSCR_DEFAULT | SR_SUBSCR_ENABLED | SR_SUBSCR_NO_THREAD,
		&module->sr_subscription);
Is it correct to use SR_SUBSCR_ENABLED in this case?   The running datastore is populated with data and as we're subscribing the callback is getting called (and failing). Seems chaotic.

**Comment by jktjkt** [comment]
All of the "real" `sr_module_change_subscribe` callbacks that I wrote and that are interested in the `running` DS, all of them use this flag. It's very common to write your callbacks in such a way that your app/plugin/whatever starts with an "empty state", and that means that there's no conceptual difference between a manual NETCONF `<edit-config>` and the initial automatic copy of config from `startup` to `running`.

But as Michal said, please don't ask about fragments. Share the big picture and explain what your goals are and what your code is doing to implement them.

**Comment by TeSwen** [comment]
So the big picture is not all that big.  I'm working with  Free Range Router code which I did not write but appears to be well written by knowledgeable folks.   I can't contact the developers as they have moved on.    Basically the code connects to the sysrepo database, creates a session and starts subscribing to the modules found in the running datastore.   I have populated the running datastore from a hand created xml file which I imported with sysrepocfg plus some edit-configs.   But as to the big picture, the goal is to successfully subscribe to these modules so that when a value is changed with edit-config, the FRR is notified by callback and updates it's router configuration with the changed value.   The problem encountered is when a module is subscribed to the callback is called right away and the error shown is that it can't find the 'node' with the xpath its gotten from sysrepo.   I've traced to this libyang function returning a null value:   lys_find_path().  So here's an xpath that it tries to find a 'node' for:   "/frr-interface:lib/interface[name='tunl0101g0']".  This is a valid path and should work.  Its all visible with 'get-config --source running'.  So trying to understand what might be wrong with the data or the path.    This explanation has been high-level so let me know what details would be helpful, for example function calls being made, yang modules loaded, the data,  etc.

**Comment by michalvasko** [comment]
So you are trying to fix a code you have not written and you do not understand? Hence my confusion, I think. Not sure I can help because while you did provide the bigger picture, it seems it is not that relevant and I am still missing information to help. You have suddenly mentioned getting the `priv` pointer value from a schema node, which is empty. I have no idea why it is being accessed and what is expected to be there, like I said, this pointer is managed by whatever application is using *libyang* (or a flag can be used for *libyang* to populate it but obviously it was not) and so I probably cannot help you no matter what.

**Comment by jktjkt** [comment]
Can you please post a link to the source code which tries to find the XPath `/frr-interface:lib/interface[name='tunl0101g0']` and fails?

**Comment by TeSwen** [comment]
Thanks Jan.   Attached are the relevant files.   The main action is in northbound_sysrepo.c.   The specific function where I'm experiencing the failure is in frr_sr_process_change.   It calls nb_node_find_with_sysrepo_context which returns null.    The function frr_sr_init is where things really start in terms of connecting and subscribing.  
 
[frr_sysrepo_issue_1975.tar.gz](https://github.com/CESNET/libyang/files/10688072/frr_sysrepo_issue_1975.tar.gz)

**Comment by jktjkt** [comment]
OK, so it seems that the original FRR code is somehow building a "shadow" tree of libyang schema nodes based on whatever (probably some internal idea of which modules are implemented). It's also heavily using the `priv` member of these schema nodes for mapping changed/accessed/whatever XPaths to some FRR-internal data structures. Then, in all sysrepo callbacks, the code is trying to somehow map from a `sr_val_t` to this internal, completely-unrelated-to-sysrepo libyang schema node, and presumably to execute whatever needs to be executed in response to config changes. (Or I might be completely off and the code is doing something totally different). Perhaps this design is documented somewhere, and if I were you I'd start trying to understand these design decisions before trying to figure out what's wrong, because frankly, without knowing *whether* and *why* the original FRR code is indeed doing what I just described, it's unlikely that you'll get relevant answers from libyang upstream.

In light of the above, it's plausible that the original code was OK-ish. As long as you don't mix sysrepo's `ly_ctx` and FRR's internal ("native") `ly_ctx`, stuff might well be OK, and it "just" needs a big fat warning that these libyang nodes are totally unrelated to sysrepo. What you absolutely **cannot** do is trying to extract the libyang data from sysrepo's context, and trying to use them as if their `priv` data was set to some FRR-internal bits -- that's what we suggested in our previous responses, and that just won't work here.

Also, code like this is incorrect:
```c
struct nb_node {
	/* Back pointer to the libyang schema node. */
	const struct lysc_node *snode;
};
sysrepo_ctx = sr_session_acquire_context(session);
nb_node = nb_node_find_with_sysrepo_context(xpath,sysrepo_ctx);
sr_session_release_context(session); 
// ...
do_someting_with(nb_node);
```
...because you're trying storing a libyang data structure (within `nb_node`) and accessing that after its context might have been freed already. On the other hand, this all happens within a sysrepo callback (right?), and sysrepo quite obviously won't release the libyang context in the middle of a callback, so you might not get a crash right away. Still, the code is wrong.

Going back to your original issue, which was a log message `OSPF: libyang: No module connected with the prefix "frr-ospfd" found (prefix format JSON module names).` -- it seems that the FRR-internal ("native") libyang context is set up in a different way compared to sysrepo's libyang context. The original message probably means that FRR does not load the `frr-ospfd` module into its internal `ly_ctx`, while the module is loaded and implemented in sysrepo. Then, once someone changes that XPath in sysrepo, FRR tries to look up that node in its internal libyang schema tree, finds nothing, and errors out. That's a FRR bug, and you'll need a FRR developer who's familiar with its architecture to debug that. There's no obvious libyang bug.
